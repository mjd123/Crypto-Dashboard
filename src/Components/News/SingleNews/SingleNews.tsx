import { Card } from "antd";
import { motion } from "framer-motion";
import { useEffect } from "react";
import { useState } from "react";
import useFetchData from "../../../hooks/useFetchData";
import { Text } from "../styles";

interface SingleNewsProps {
  news:
    | {
        title?: string;
        url?: string;
        urlToImage?: string;
        description?: string;
      }
    | never[];
  id?: number;
}

const variants = {
  fadeIn: { opacity: 1 },
  fadeOut: { opacity: 0 },
};

const SingleNews = ({ news, id }: SingleNewsProps) => {
  //const [newsItems, setNewsItems] = useState<any>([]);
  const [currentNewsItem, setCurrentNewsItem] = useState(0);
  const [fadeComponent, setFadeComponent] = useState(false);

  const [fetchTrigger, setFetchTrigger] = useState<boolean>();
  const { newsData } = useFetchData(fetchTrigger);
  let newsItems: any = news;

  useEffect(() => {
    if (id === 0 && currentNewsItem === newsItems.length - 1) {
      setFetchTrigger(true);
    } else {
      setFetchTrigger(false);
    }
  }, [news]);

  return (
    <>
      {newsItems && (
        <motion.a
          initial={{ opacity: 1 }}
          animate={fadeComponent ? "fadeOut" : "fadeIn"}
          variants={variants}
          transition={{
            ease: "easeInOut",
            duration: 1,
            delay: fadeComponent ? 45 : 0.5,
          }}
          href={newsItems[currentNewsItem]?.url}
          target="_blank"
          rel="noreferrer"
          style={{ display: "flex", height: "100%" }}
          onAnimationComplete={(definition) => {
            setFadeComponent(true);
            if (definition === "fadeOut") {
              setFadeComponent(false);
              setTimeout(() => {
                setCurrentNewsItem(
                  currentNewsItem < newsItems.length - 1
                    ? currentNewsItem + 1
                    : 0
                );
              }, 10);
            }
          }}
        >
          <Card
            bordered={false}
            hoverable={true}
            bodyStyle={{
              position: newsItems[currentNewsItem]?.urlToImage
                ? "absolute"
                : "initial",
              display: "flex",
              flexDirection: "column",
              justifyContent: "space-around",
              alignItems: "center",
              textAlign: "center",
              height: "100%",
              minHeight: "50px",
              boxShadow: "inset 0 0 0 100vmax rgba(0, 0, 0, 0.6)",
              borderRadius: "10px",
            }}
            size="small"
            style={{
              display: "flex",
              flexDirection: "column",
              width: "100%",
              borderRadius: "10px",
            }}
            cover={
              newsItems[currentNewsItem]?.urlToImage && (
                <img
                  style={{ borderRadius: "10px" }}
                  alt={newsItems[currentNewsItem].description}
                  src={newsItems[currentNewsItem].urlToImage}
                />
              )
            }
          >
            <Text color="white">{newsItems[currentNewsItem]?.title}</Text>
          </Card>
        </motion.a>
      )}
    </>
  );
};

export default SingleNews;
