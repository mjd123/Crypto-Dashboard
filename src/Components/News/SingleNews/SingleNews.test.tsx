import React from "react";
import { shallow, mount } from "enzyme";
import SingleNews from "./SingleNews";
import toJson from "enzyme-to-json";

const mockData = {
  news: [
    {
      title: "title",
      url: "url",
      description: "description",
      urlToImage: "urlToImage",
      content: "content",
    },
  ],
};

describe("component renders", () => {
  test("SingleNews renders without crashing", () => {
    const component = shallow(<SingleNews news={mockData.news} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});

describe("test props", () => {
  test("SingleNews renders props", () => {
    const component = mount(<SingleNews news={mockData.news} />);
    const value = component.find("span").text();
    //console.log(value, mockData.news[0].title);

    expect(value).toEqual(mockData.news[0].title);
  });

  test("SingleNews renders without props", () => {
    const component = mount(<SingleNews news={[{}]} />);

    expect(toJson(component)).toMatchSnapshot();
  });

  test("check the type of value", () => {
    const component = mount(<SingleNews news={mockData.news} />);
    expect(component.props().news[0].title).toMatch("title");
  });
});
