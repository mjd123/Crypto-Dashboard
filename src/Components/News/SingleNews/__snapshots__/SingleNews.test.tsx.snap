// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders SingleNews renders without crashing 1`] = `
<Fragment>
  <a
    rel="noreferrer"
    target="_blank"
  >
    <Card
      bodyStyle={
        Object {
          "alignItems": "center",
          "borderRadius": "10px",
          "boxShadow": "inset 0 0 0 100vmax rgba(0, 0, 0, 0.6)",
          "display": "flex",
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "space-around",
          "position": "initial",
          "textAlign": "center",
        }
      }
      bordered={false}
      hoverable={true}
      size="small"
      style={
        Object {
          "borderRadius": "10px",
          "display": "flex",
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <Styled(PageText)
        color="white"
      />
    </Card>
  </a>
</Fragment>
`;

exports[`test props SingleNews renders without props 1`] = `
<SingleNews
  news={
    Array [
      Object {},
    ]
  }
>
  <a
    rel="noreferrer"
    target="_blank"
  >
    <Card
      bodyStyle={
        Object {
          "alignItems": "center",
          "borderRadius": "10px",
          "boxShadow": "inset 0 0 0 100vmax rgba(0, 0, 0, 0.6)",
          "display": "flex",
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "space-around",
          "position": "initial",
          "textAlign": "center",
        }
      }
      bordered={false}
      hoverable={true}
      size="small"
      style={
        Object {
          "borderRadius": "10px",
          "display": "flex",
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <div
        className="ant-card ant-card-hoverable ant-card-small"
        style={
          Object {
            "borderRadius": "10px",
            "display": "flex",
            "flexDirection": "column",
            "width": "100%",
          }
        }
      >
        <div
          className="ant-card-body"
          style={
            Object {
              "alignItems": "center",
              "borderRadius": "10px",
              "boxShadow": "inset 0 0 0 100vmax rgba(0, 0, 0, 0.6)",
              "display": "flex",
              "flexDirection": "column",
              "height": "100%",
              "justifyContent": "space-around",
              "position": "initial",
              "textAlign": "center",
            }
          }
        >
          <Styled(PageText)
            color="white"
          >
            <PageText
              className="sc-bdfBQB jaKmxc"
              color="white"
            >
              <span
                className="sc-bdfBQB jaKmxc"
              />
            </PageText>
          </Styled(PageText)>
        </div>
      </div>
    </Card>
  </a>
</SingleNews>
`;
